blog post about testing CSI drivers

pohly opened the pull request:
The earlier E2E blog post touch on some of these aspects, but was more
focused on writing a custom test suite. Using the existing tests
unchanged may be simpler and sufficient. Because it is not that
well-known yet, a blog post seems useful.

pohly commented:
/cc @msau42 

This is WIP because some of the components described in the blog post haven't been released yet. I think it makes sense to merge https://github.com/kubernetes-csi/csi-lib-utils/pull/34 first and do csi-test 2.3.0 and csi-driver-host-path 1.2.0 releases first.

It's also a very early draft - still need to go through it myself again tomorrow.

msau42 commented on content/en/blog/_posts/2019-10-15-csi-driver-testing.md:
```diff
@@ -0,0 +1,397 @@
+
+[csi-test
+sanity](https://github.com/kubernetes-csi/csi-test/tree/master/pkg/sanity)
+ensure that a CSI driver conforms to the CSI specification. Despite
```
typo: ensures

pohly replied:
Fixed.

msau42 commented on content/en/blog/_posts/2019-10-15-csi-driver-testing.md:
````diff
@@ -0,0 +1,397 @@
+EOF
+```
+
+The `FromName` fields tell the tests to create storage and snapshot
````
Add a note that it creates a StorageClass with no parameters. If you driver needs parameters specified, then it is better to use the FromFile option.

pohly replied:
I've removed this entirely and instead refer the reader to the comments in the source code.

pohly commented:
Michelle Au <notifications@github.com> writes:
> I think some of the content here may be a little too detailed for a
> blog post, but very well suited in the README.

I'm worried that putting too much text into the README will become a
liability. We would have to ensure that it is always up-to-date. We
already failed with that: the README.md still uses `dataSource`, but
https://github.com/kubernetes/kubernetes/pull/80117/files#diff-84aa234538402d584aee76f4923fcf8f
replaced that with `pvcDataSource`.

A blog post of course doesn't solve that problem. But at least we don't
disappoint readers because the expectation will be that the instructions
work for the versions of the software that is described in it, and that
I have checked.

But suppose we copy text into the README: should I then copy the entire
"End-to-end testing" section into the README and only keep a simplified
version in the blog post?

> Also maybe emphasizing more on the benefits to CSI driver developers,
> and a get involved section at the end.

Okay.

> > FromName
> Add a note that it creates a StorageClass with no parameters. If you
> driver needs parameters specified, then it is better to use the
> FromFile option.

Okay.

msau42 commented:
I think details about things like how to set specific fields should live with the code. Maybe documenting it in the struct and then referencing the file from the README is sufficient.

pohly commented:
Michelle Au <notifications@github.com> writes:
> I think details about things like how to set specific fields should
> live with the code. Maybe documenting it in the struct and then
> referencing the file from the README is sufficient.

That's also what I would prefer. In the README.md it's easy to reference
the code, too.

For the blog post I was trying to pull that information together for a
specific example that actually works. If that's too detailed and/or too
long, then I can shorten it.

pohly commented:
I've pushed an updated version.

/assign @msau42 

@msau42: can you have another look?

@timoreimann : you said on Slack that you also wanted to check it out - this is your chance! :grin:

timoreimann commented:
Thanks @pohly, will have a look.

Re: leaving documentation in the code and pointing to it from a README: I understand the desire to avoid documentation from becoming outdated (who doesn't), but I feel there are a few issues with that approach:

1. If you don't use a permalink, you risk pointing at a file that stops to exist after the next refactoring (of which there are generally plenty in Kubernetes).
1. If you use a permalink, you risk not pointing at the latest version.
1. Unrendered documentation is fairly hard to read, and the capabilities of ASCII to format are limited.

Personally, I favor to have instructions in a README file even if it means they could be out of date at times.

An alternative / addition could be to have executable examples, e.g., something in Go or bash that can be validated during CI.

timoreimann commented on content/en/blog/_posts/2019-10-15-csi-driver-testing.md:
```diff
@@ -0,0 +1,317 @@
+issue trackers select issues that have been marked specifically as
+something that needs someone's help:
+- [csi-test](https://github.com/kubernetes-csi/csi-test/issues?q=is%3Aissue+is%3Aopen+label%3A%22help+wanted%22)
+- [Kubernetes](https://github.com/kubernetes/kubernetes/issues?utf8=%E2%9C%93&q=is%3Aopen+is%3Aissue+label%3A%22help+wanted%22+label%3Asig%2Fstorage+)
```
Some final closing words after this one might be nice.

pohly replied:
Suggestions? All I could come up with is:

> Happy testing! May the issues it find be few and easy to fix.

timoreimann replied:
I like that closing a lot. 😄 🚢

pohly replied:
It was still stuck in an unsaved editor buffer... now it's in the branch.

timoreimann commented on content/en/blog/_posts/2019-10-15-csi-driver-testing.md:
````diff
@@ -0,0 +1,317 @@
+SUCCESS! -- 53 Passed | 0 Failed | 0 Pending | 19 Skipped
+PASS
+```
+
````
Can we link (or just refer) to the implementation of the hostpath sanity tests for readers to have something to model after?

pohly replied:
These are the tests from https://github.com/kubernetes-csi/csi-test/tree/master/pkg/sanity, i.e. not some tests that are specific to hostpath.

I added:
* The source code of these tests is in the [`pkg/sanity`
  package](https://github.com/kubernetes-csi/csi-test/tree/master/pkg/sanity).

pohly commented:
> Re: leaving documentation in the code and pointing to it from a README: I understand the desire to avoid documentation from becoming outdated (who doesn't), but I feel there are a few issues with that approach:
> ...

The README.md is https://github.com/kubernetes/kubernetes/blob/master/test/e2e/storage/external/README.md. This can use a relative link, so there's no risk of getting the wrong or out-dated source code. It also gets rendered by GitHub, so your points should be addressed.

However, it did get outdated when changes were made to the source code without updating that documentation. That's why I want to avoid putting too much content into it.

> Personally, I favor to have instructions in a README file even if it means they could be out of date at times.

As a user I prefer a README, but not if it is out-dated and thus potentially misleading. As developer, duplicating information becomes a burden. :grimacing:

pohly commented:
Updated in-tree documentation: https://github.com/kubernetes/kubernetes/pull/86077

pohly commented on content/en/blog/_posts/2019-10-15-csi-driver-testing.md:
```diff
@@ -0,0 +1,317 @@
+from remote by forwarding incoming TCP connections with
+[`socat`]((https://github.com/intel/pmem-csi/blob/82e6114a7282015cabc0fdcd46c24776bf3876dd/deploy/kubernetes-1.14/pmem-csi-direct-testing.yaml#L221-L229).
+
+The Kubernetes-CSI example driver also gets tested on Kubernetes this
```
Or better "CSI hostpath example driver" - there's also an in-tree hostpath driver.

pohly commented on content/en/blog/_posts/2019-10-15-csi-driver-testing.md:
```diff
@@ -0,0 +1,317 @@
+
+$ chmod u+x *_in_pod.sh
+$ csi-sanity -ginkgo.v \
+             -csi.junitfile junit_sanity.xml \
```
Except that https://github.com/kubernetes-csi/csi-test/tree/master/cmd/csi-sanity/README.md doesn't mention it either and just refers to `csi-sanity -h`. Maintenance effort again (and this time it wasn't me who created that README.md!).

I wanted to show-case that the option exists. But as I don't want make the post longer by explaining it I removed the line.

pohly commented on content/en/blog/_posts/2019-10-15-csi-driver-testing.md:
```diff
@@ -0,0 +1,317 @@
+
+Some comments:
+
+* How to determine the external IP address of the node depends on the
```
Changed into
> In this example, the cluster was brought up with
  `hack/local-up-cluster.sh` and thus runs on the local host (`127.0.0.1`).
  It uses a port allocated by Kubernetes, obtained above with `kubectl get "services/hostpath-service"`.

pohly commented on content/en/blog/_posts/2019-10-15-csi-driver-testing.md:
```diff
@@ -0,0 +1,317 @@
+  Docker
+  command](https://github.com/kubernetes-csi/csi-driver-host-path/blob/3488dc7f994e33485629b86b69a6f34ebb7ef2d9/release-tools/prow.sh#L850)
+  can be used.
+* There's no need to have separate scripts for staging and mount
```
Fine with me. Removed.

pohly commented on content/en/blog/_posts/2019-10-15-csi-driver-testing.md:
```diff
@@ -0,0 +1,317 @@
+### Installation
+
+For each Kubernetes release, a test tar archive is published. It's not
+listed in the release notes (for example, the ones for
```
> For each Kubernetes release, a test tar archive is published. It's not
listed in the release notes (for example, the ones for
[1.16](https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG-1.16.md#downloads-for-v1161)),
so one has to know that the full URL is
`https://dl.k8s.io/<version>/kubernetes-test-linux-amd64.tar.gz` (like
for
[v1.16.0](https://dl.k8s.io/v1.16.0/kubernetes-test-linux-amd64.tar.gz)).

sftim replied:
Maybe update the examples for v1.17.0?

pohly commented on content/en/blog/_posts/2019-10-15-csi-driver-testing.md:
````diff
@@ -0,0 +1,317 @@
+[`csi-driver-host-path`](https://github.com/kubernetes-csi/csi-driver-host-path):
+
+```
+$ cat >test-driver.yaml <<EOF
````
It was removed. Fix for documentation pending in https://github.com/kubernetes/kubernetes/pull/86077.

pohly commented on content/en/blog/_posts/2019-10-15-csi-driver-testing.md:
````diff
@@ -0,0 +1,317 @@
+
+```
+$ ./e2e.test -ginkgo.v \
+             -ginkgo.focus='External.Storage.*hostpath.csi.k8s.io' \
````
It's enough to focus on just `External.Storage`. Adding the driver name may be useful when multiple external drivers are enabled - but then one can also just select the driver by using only its `-storage.testdriver`.

I've simplified it.

pohly commented on content/en/blog/_posts/2019-10-15-csi-driver-testing.md:
````diff
@@ -0,0 +1,317 @@
+```
+
+If the [`ginkgo`
+command](https://onsi.github.io/ginkgo/#getting-ginkgo) is installed,
````
Actually, yes, it seems to be in `kubernetes/test/bin/ginkgo`.

I've extended the unpack instructions above and just use the command here.

pohly commented on content/en/blog/_posts/2019-10-15-csi-driver-testing.md:
```diff
@@ -0,0 +1,317 @@
+If the [`ginkgo`
+command](https://onsi.github.io/ginkgo/#getting-ginkgo) is installed,
+then one can also run tests in parallel, except for alpha feature
+tests or those that by design have to be run sequentially:
```
> With the `ginkgo` command one can also run some tests in parallel.
Alpha feature tests or those that by design have to be run
sequentially then need to be run separately:

...

```
$ ./ginkgo -p -v \
         -focus='External.Storage.*(\[Feature:|\[Disruptive\])' \
         ./e2e.test \
         -- \
         -storage.testdriver=test-driver.yaml
```

pohly commented on content/en/blog/_posts/2019-10-15-csi-driver-testing.md:
```diff
@@ -0,0 +1,317 @@
+although it complies with the CSI specification. If that happens then
+please file issues (links below).
+
+But these are open source projects which depend on the help of those
```
Done.

I wanted to stress that filing an issue is no guarantee that someone will solve it.

pohly commented:
> I left a couple of remarks, made from the perspective of a fairly new user to the test suite.

Thanks, that's exactly the kind of feedback that we need.

pohly commented:
I pushed the updated version.

msau42 commented:
/lgtm
Thanks!

pohly commented:
/assign @castrojo

Content is ready, we just need to select a publishing date and update the date in the text.

timoreimann commented on content/en/blog/_posts/2019-10-15-csi-driver-testing.md:
```diff
@@ -0,0 +1,324 @@
+         ./e2e.test \
+         -- \
+         -storage.testdriver=test-driver.yaml
+$ ./ginkgo -p -v \
```
Doesn't this has has to go without `-p` since it's executing the not-to-be-parallelized feature and disruptive tests?

pohly replied:
Good catch. Fixed.

pohly commented on content/en/blog/_posts/2019-10-15-csi-driver-testing.md:
```diff
@@ -0,0 +1,324 @@
+**Author:** Patrick Ohly (Intel)
+
+When developing a [Container Storage Interface (CSI)
+driver](https://kubernetes-csi.github.io/docs/) is useful to leverage
```
There's also a missing "it" here - will fix.

sftim replied:
🤦 yes!

pohly replied:
Done.

sftim commented:
I forgot to mention this was informal feedback. I'm not able to help out with a scheduling date.
@kubernetes/sig-docs-blog-reviewers can help here I think.

koreno commented:
As a new CSI developer, I just landed on this and it seems incredibly useful.
However, I think there is still a gap - there should more information/instruction on how to get the e2e tests to run with some custom CSI plugin, perhaps in the form of a step-by-step tutorial.

The main questions that came up -
* should I deploy the plugin manually on my k8s cluster and run then run the tests? or do the tests do the deployment?
* what does each flag/option in the `test-driver.yaml` mean? (the comments in the associated structs didn't help much; the text is a bit laconic/brief)

From my understanding e2e.test is the to test many other k8s features, not just CSI. This document should make it straightforward for a CSI developer to use e2e.test, someone who might know very little about this testing framework and many k8s-specifc details.

koreno commented on content/en/blog/_posts/2019-10-15-csi-driver-testing.md:
````diff
@@ -0,0 +1,324 @@
+```
+$ cat >test-driver.yaml <<EOF
+StorageClass:
+  FromName: true
````
What about the alternative and when/how should I use it?
It would be helpful to clarify

pohly replied:
Looks like we are going in circles here :cry: I had that in the original text and @msau42 wanted me to make the blog post shorter.

Let's keep it as is, which means you'll have to follow the links and read the documentation for each field. There you'll find:
https://github.com/kubernetes/kubernetes/blob/2bd9643cee5b3b3a5ecbd3af49d09018f0773c77/test/e2e/storage/external/external.go#L168-L170

koreno replied:
okay, then maybe just make sure there's a link or some reasonable way for finding that comment block.

msau42 replied:
I preferred the blog post to give a simple example, and then point to the documentation that explains in more detail all of the fields.

You can add a summary like "at a minimum, you need to define the storageclass and snapshotclasses you want to use in the test, the name of your driver, and what capabilities you want to test"

pohly replied:
> point to the documentation that explains in more detail all of the fields

@koreno : that's what the `DriverInfo` and `driverDefinition` links are for. If they aren't sufficient, we need to enhance them; this blog post isn't the right place for providing documentation.

@msau42: I added your summary. I left out "snapshotclass" because it's optional.

pohly commented on content/en/blog/_posts/2019-10-15-csi-driver-testing.md:
```diff
@@ -0,0 +1,324 @@
+
+Starting with csi-test v3.0.0, the `csi-sanity` command can be built
+with `go get github.com/kubernetes-csi/csi-test/cmd/csi-sanity` and
+then can be found in `$GOPATH/bin/csi-sanity`.
```
Done.

pohly commented on content/en/blog/_posts/2019-10-15-csi-driver-testing.md:
```diff
@@ -0,0 +1,324 @@
+that is embedded inside it. For detailed usage instructions of
+individual fields refer to the source code of the `e2e.test` binary.
+
+WARNING: tests are often only run when setting some fields and the
```
Done.

pohly commented:
> As a new CSI developer, I just landed on this and it seems incredibly useful.
However, I think there is still a gap - there should more information/instruction on how to get the e2e tests to run with some custom CSI plugin, perhaps in the form of a step-by-step tutorial.

But isn't that what the blog post provides, using the hostpath driver as example for such a custom CSI plugin?

>    should I deploy the plugin manually on my k8s cluster and run then run the tests? or do the tests do the deployment?

That part wasn't clear. It must be started manually. I've changed two sentences and added the `deploy-hostpath.sh` invocation to address this.

>    what does each flag/option in the test-driver.yaml mean? (the comments in the associated structs didn't help much; the text is a bit laconic/brief)

I fear this is where we need specific feedback and improvement suggestions. Perhaps you can take notes while going through the options and write down what's missing, then file an issue or PR against the Kubernetes repo?

For a more experience developer it may not be obvious what's missing because even a brief comment might be enough.

koreno commented:
> > As a new CSI developer, I just landed on this and it seems incredibly useful.
> > However, I think there is still a gap - there should more information/instruction on how to get the e2e tests to run with some custom CSI plugin, perhaps in the form of a step-by-step tutorial.
> 
> But isn't that what the blog post provides, using the hostpath driver as example for such a custom CSI plugin?
> 
It requires a developer to know the hostpath implementation so that one can compare their own CSI plugin to the hostpath, and decide how to fill the various fields. But this might be good enough.

Also notice that the suggested command line includes a regex filter ("focus") that includes "hostpath", which could indicate that the selected tests might not cover all necessary features, or actually run tests that are specific to the hostpath driver.

> > should I deploy the plugin manually on my k8s cluster and run then run the tests? or do the tests do the deployment?
> 
> That part wasn't clear. It must be started manually. I've changed two sentences and added the `deploy-hostpath.sh` invocation to address this.
> 
> > what does each flag/option in the test-driver.yaml mean? (the comments in the associated structs didn't help much; the text is a bit laconic/brief)
> 
> I fear this is where we need specific feedback and improvement suggestions. Perhaps you can take notes while going through the options and write down what's missing, then file an issue or PR against the Kubernetes repo?
> 
> For a more experience developer it may not be obvious what's missing because even a brief comment might be enough.

msau42 commented on content/en/blog/_posts/2019-10-15-csi-driver-testing.md:
```diff
@@ -0,0 +1,348 @@
+and [the DriverInfo
+struct](https://github.com/kubernetes/kubernetes/blob/v1.16.0/test/e2e/storage/testsuites/testdriver.go#L139-L185)
+that is embedded inside it. For detailed usage instructions of
+individual fields refer to the source code of the `e2e.test` binary.
```
link to the source code?

msau42 commented on content/en/blog/_posts/2019-10-15-csi-driver-testing.md:
````diff
@@ -0,0 +1,348 @@
+```
+$ ./ginkgo -p -v \
+         -focus='External.Storage' \
+         -skip='\[Feature:|\[Disruptive\]' \
````
skip should also include "[Serial]"

pohly replied:
Fixed.

msau42 commented:
/lgtm
/hold
for updating the blog date

pohly commented:
@castrojo: any suggestions for when this could or should be published?

Can you update the date as part of the approval or do I need to do that? It also better should get squashed.

mrbobbytables commented:
Everything looks good from the technical side 👍 
/assign @kbarnard10 
for final review

sftim commented on content/en/blog/_posts/2019-10-15-csi-driver-testing.md:
````diff
@@ -0,0 +1,351 @@
+Here is an example that tests the
+[`csi-driver-host-path`](https://github.com/kubernetes-csi/csi-driver-host-path):
+
+```
````
Optionally: mark this as YAML and remove the first / last lines, then name the file in the text.
(In documentation, we'd make the file downloadable using a `{{< codenew >}}` shortcode).

Reason for marking as YAML is that it's more aesthetic and therefore approachable.
An alternative: remove the initial `$ ` and mark it as Bourne-flavored shell.

sftim commented on content/en/blog/_posts/2019-10-15-csi-driver-testing.md:
````diff
@@ -0,0 +1,351 @@
+`-ginkgo.focus`:
+
+```
+$ ./e2e.test -ginkgo.v \
````
In the main documentation it's usual to [separate commands from their output](https://kubernetes.io/docs/contribute/style/style-guide/#separate-commands-from-output). Whilst that's not required for blog posts, you might want to adopt the same style.

sftim commented on content/en/blog/_posts/2019-10-15-csi-driver-testing.md:
````diff
@@ -0,0 +1,351 @@
+...
+```
+
+With the `ginkgo` command one can also run some tests in parallel.
````
Maybe:
```suggestion
You can use `ginkgo` to run some kinds of test in parallel.
```

sftim commented on content/en/blog/_posts/2019-10-15-csi-driver-testing.md:
```diff
@@ -0,0 +1,351 @@
+Kubernetes API. Therefore a configuration file in YAML or JSON format
+is needed which describes the driver that is to be tested. That file
+is used to populate [the driverDefinition
+struct](https://github.com/kubernetes/kubernetes/blob/v1.16.0/test/e2e/storage/external/external.go#L142-L211)
```
Update these links for v1.17?

sftim commented on content/en/blog/_posts/2019-10-15-csi-driver-testing.md:
```diff
@@ -0,0 +1,351 @@
+In contrast to `csi-sanity`, end-to-end testing interacts with the CSI
+driver through the Kubernetes API, i.e. it simulates operations from a
+normal user, like creating a PVC. Support for testing external CSI
+drivers [was
```
nit: I'd have the link text be _added_ (move “was” out of the link)

sftim commented on content/en/blog/_posts/2019-10-15-csi-driver-testing.md:
```diff
@@ -0,0 +1,351 @@
+
+In contrast to `csi-sanity`, end-to-end testing interacts with the CSI
+driver through the Kubernetes API, i.e. it simulates operations from a
+normal user, like creating a PVC. Support for testing external CSI
```
In case readers aren't up to speed on Kubernetes storage concepts, how about:
```suggestion
normal user, like creating a PersistentVolumeClaim. Support for testing external CSI
```

sftim commented on content/en/blog/_posts/2019-10-15-csi-driver-testing.md:
````diff
@@ -0,0 +1,351 @@
+
+Here's how one can replicate that, using the v1.2.0 release:
+
+```
````
In the main documentation it's usual to [separate commands from their output](https://kubernetes.io/docs/contribute/style/style-guide/#separate-commands-from-output). Whilst that's not required for blog posts, you might want to adopt the same style.

sftim commented on content/en/blog/_posts/2019-10-15-csi-driver-testing.md:
```diff
@@ -0,0 +1,351 @@
+[Scripts](https://github.com/kubernetes-csi/csi-driver-host-path/blob/v1.2.0/release-tools/prow.sh#L808-L859)
+are used to create and remove the directories.
+
+Here's how one can replicate that, using the v1.2.0 release:
```
v1.2.0 of Kubernetes? Is it worth being super clear here, in case a small number of readers get confused?

sftim commented on content/en/blog/_posts/2019-10-15-csi-driver-testing.md:
```diff
@@ -0,0 +1,351 @@
+written in Go, the driver itself can be implemented in any language.
+
+The main
+[README.md](https://github.com/kubernetes-csi/csi-test/blob/master/pkg/sanity/README.md)
```
```suggestion
[README](https://github.com/kubernetes-csi/csi-test/blob/master/pkg/sanity/README.md)
```

pohly commented:
> @pohly here's some feedback. It's almost all optional stuff rather than you-must-fix-before-publication.

These are all useful suggestions. But if there are no other changes required, I would prefer to get this published as-is.

castrojo commented:
/lgtm

Just needs the date metadata fixed.

pohly commented:
> Just needs the date metadata fixed.

In that case let me propose that we publish this week on Wednesday. I don't have any strong reasons for this date, though, so just let me know if some other day is better.

In addition, I've squashed the PR commits and addressed the minor, one-line (or one-word) changes that @sftim suggested.

castrojo commented:
/lgtm

Thanks!

pohly commented:
/assign @sig-docs-blog-owners

For final approval.

sftim commented:
@kubernetes/sig-docs-blog-owners ?

castrojo commented:
/hold cancel
/approve

Thanks, sorry that you got stuck in the holiday-delay!
